risc0/zkvm/src/host/client/env.rs

```diff
@@ -75,6 +66,29 @@ impl SegmentPath {
     }
 }
 
+/// A ZKR proof request.
+#[stability::unstable]
+pub struct ProveZkrRequest {
+    /// The digest of the claim that this ZKR program is expected to produce.
+    pub claim_digest: Digest,
+
+    /// The control ID uniquely identifies the ZKR program to be proven.
+    pub control_id: Digest,
+
+    /// The input that the ZKR program should operate on.
+    pub input: Vec<u8>,
+}
+
+/// A trait that supports the ability to be notified of ZKR proof requests
+/// on-demand.
+#[stability::unstable]
+pub trait CoprocessorCallback {
+    /// Request that a proof of a ZKR is produced.
+    fn prove_zkr(&mut self, request: ProveZkrRequest) -> Result<()>;
+}
+
+pub type CoprocessorCallbackRef<'a> = Rc<RefCell<dyn CoprocessorCallback + 'a>>;
```

Comment:
Another option here would be to make `prove_zkr` take an immutable `&self` and let the implementer chose to use a `Cell` or `RefCell` if they need interior mutability. It's a matter of whether or not the callback semantically carries state. Given the return type, I imagine the answer is "yes", so maybe this signature is best anyways.

Reply:
I've found it's much easier to just assume that the handler will mutate state.  But I've gone back and forth on this and I think this is consistent with other callbacks we have.